File: ios/Classes/Map4dServicesPlugin.m
#import "Map4dServicesPlugin.h"
#import "SRequest.h"
#import "SClient.h"

@implementation Map4dServicesPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel methodChannelWithName:@"map4d_services"
                                                              binaryMessenger:[registrar messenger]];
  Map4dServicesPlugin* instance = [[Map4dServicesPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  
  /* Geocode */
  if ([@"geocode#geocode" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildGeocodingRequestWithData:call.arguments] result:result];
    return;
  }

  /* Place | Place Detail */
  if ([@"place#detail" isEqualToString:call.method]) {
    NSString *placeId = call.arguments[@"id"];
    [SClient fireRequest:[SRequest buildPlaceDetailRequestWithId:placeId] result:result];
    return;
  }
  
  /* Place | Text Search */
  if ([@"place#text-search" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildTextSearchRequestWithData:call.arguments] result:result];
    return;
  }
  
  /* Place | Nearby Search */
  if ([@"place#nearby-search" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildNearbySearchRequestWithData:call.arguments] result:result];
    return;
  }
  
  /* Place | Viewbox Search */
  if ([@"place#viewbox-search" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildViewboxSearchRequestWithData:call.arguments] result:result];
    return;
  }
  
  /* Place | Suggestion */
  if ([@"place#autosuggest" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildSuggestionsRequestWithData:call.arguments] result:result];
    return;
  }
  
  /* Route | Directions */
  if ([@"route#route" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildDirectionsRequestWithData:call.arguments] result:result];
    return;
  }
  
  /* Route | ETA */
  if ([@"route#eta" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildRouteETARequestWithData:call.arguments] result:result];
    return;
  }
  
  /* Route | Matrix */
  if ([@"route#matrix" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildDistanceMatrixRequestWithData:call.arguments] result:result];
    return;
  }
  
  /* Route | Graph */
  if ([@"route#graph" isEqualToString:call.method]) {
    [SClient fireRequest:[SRequest buildGraphRouteRequestWithData:call.arguments] result:result];
    return;
  }
  
  NSLog(@"[Map4dServicesPlugin] Method not implemented: %@", call.method);
  result(FlutterMethodNotImplemented);
}

@end

File: android/src/main/java/vn/map4d/services/map4d_services/Map4dServicesPlugin.java
package vn.map4d.services.map4d_services;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import vn.map4d.services.MFServicesOptions;

/** Map4dServicesPlugin */
public class Map4dServicesPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  private MFServicesOptions servicesOptions;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "map4d_services");
    channel.setMethodCallHandler(this);
    servicesOptions = new MFServicesOptions.Builder(flutterPluginBinding.getApplicationContext()).build();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method) {
      case "geocode#geocode": {
        SClient.fireRequest(servicesOptions, SRequest.buildGeocodingRequestWithData(call), result);
        break;
      }
      case "place#detail": {
        SClient.fireRequest(servicesOptions, SRequest.buildPlaceDetailRequestWithId(call), result);
        break;
      }
      case "place#text-search": {
        SClient.fireRequest(servicesOptions, SRequest.buildTextSearchRequestWithData(call), result);
        break;
      }
      case "place#nearby-search": {
        SClient.fireRequest(servicesOptions, SRequest.buildNearbySearchRequestWithData(call), result);
        break;
      }
      case "place#viewbox-search": {
        SClient.fireRequest(servicesOptions, SRequest.buildViewboxSearchRequestWithData(call), result);
        break;
      }
      case "place#autosuggest": {
        SClient.fireRequest(servicesOptions, SRequest.buildSuggestionsRequestWithData(call), result);
        break;
      }
      case "route#route": {
        SClient.fireRequest(servicesOptions, SRequest.buildDirectionsRequestWithData(call), result);
        break;
      }
      case "route#eta": {
        SClient.fireRequest(servicesOptions, SRequest.buildRouteETARequestWithData(call), result);
        break;
      }
      case "route#matrix": {
        SClient.fireRequest(servicesOptions, SRequest.buildDistanceMatrixRequestWithData(call), result);
        break;
      }
      case "route#graph": {
        SClient.fireRequest(servicesOptions, SRequest.buildGraphRouteRequestWithData(call), result);
        break;
      }
      default:
        result.notImplemented();
        break;
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}
